07-routing-gossip.md

```diff
@@ -510,16 +521,196 @@ An endpoint node:
+  - MAY send this if it receives a `channel_update` for a
+   `short_channel_id` for which has no `channel_announcement`.
+  - SHOULD NOT send this if the channel referred to is not an unspent output.
+
+The receiver:
+  - if the first byte of `encoded_short_ids` is not a known encoding type:
+    - MAY fail the connection
+  - if `encoded_short_ids` does not decode into a whole number of `short_channel_id`:
+    - MAY fail the connection.
+  - if it has not sent `reply_short_channel_ids_end` to a previously received `query_short_channel_ids` from this sender:
+    - MAY fail the connection.
+  - MUST respond to each known `short_channel_id` with a `channel_announcement`
+    and the latest `channel_update`s for each end
+	- SHOULD NOT wait for the next outgoing gossip flush to send these.
+  - MUST follow with any `node_announcement`s for each `channel_announcement`
+	- SHOULD avoid sending duplicate `node_announcements` in response to a single `query_short_channel_ids`.
+  - MUST follow these responses with `reply_short_channel_ids_end`.
+  - if does not maintain up-to-date channel information for `chain_hash`:
+	- MUST set `complete` to 0.
+  - otherwise:
+	- SHOULD set `complete` to 1.
+
+#### Rationale
+
+Future nodes may not have complete information; they certainly won't have
+complete information on unknown `chain_hash` chains.  While this `complete`
+field cannot be trusted, a 0 does indicate that the sender should search
+elsewhere for additional data.
+
+The explicit `reply_short_channel_ids_end` message means that the receiver can
+indicate it doesn't know anything, and the sender doesn't need to rely on
+timeouts.  It also causes a natural ratelimiting of queries.
+
+### The `query_channel_range` and `reply_channel_range` Messages
+
+1. type: 263 (`query_channel_range`) (`gossip_queries`)
+2. data:
+    * [`32`:`chain_hash`]
+    * [`4`:`first_blocknum`]
+    * [`4`:`number_of_blocks`]
+
+1. type: 264 (`reply_channel_range`) (`gossip_queries`)
+2. data:
+    * [`32`:`chain_hash`]
+    * [`4`:`first_blocknum`]
+    * [`4`:`number_of_blocks`]
+    * [`1`:`complete`]
+    * [`2`:`len`]
+    * [`len`:`encoded_short_ids`]
+
+This allows a query for channels within specific blocks.
+
+#### Requirements
+
+The sender of `query_channel_range`:
+  - MUST NOT send this if it has sent a previous `query_channel_range` to this peer and not received `reply_channel_range`.
+  - MUST set `chain_hash` to the 32-byte hash that uniquely identifies the chain
+  that it wants the `reply_channel_range` to refer to
+  - MUST set `first_blocknum` to the first block it wants to know channels for
+  - MUST set `number_of_blocks` to 1 or greater.
+
+The receiver of `query_channel_range`:
+  - if it has not sent `reply_channel_range` to a previously received `query_channel_range` from this sender:
+    - MAY fail the connection.
+  - MUST respond with one or more `reply_channel_range` whose combined range
```

Comment:
Multiple `reply_channel_range` may make it ambiguous whether we are violating "MUST NOT send this (`query_channel_range`) if it has sent a previous `query_channel_range` to this peer and not received `reply_channel_range`" requirement.

In particular if there are any data or message buffers between the operating code of the peers (e.g. c-lightning, where sending of messages is buffered in a message queue):

1.  Peer A sends a `query_channel_range`.
2.  Peer B sends two `reply_channel_range` in response.
3.  Peer A receives one `reply_channel_range` and thinks it has fulfilled the above requirement for the first `query_channel_range`.
4.  Peer A sends another `query_channel_range`.
5.  Peer B receives the above message, then gets stalled in processing due to OS scheduling or etc.
6.  Peer A receives the second `reply_channel_range` and thinks it has fulfilled the above requirement for the second `query_channel_range`.
7.  Peer A sends a third `query_channel_range`.
8.  Peer B is awakened by the OS again, and sees another `query_channel_range` even though it has not finished processing the second `query_channel_range`.  It then takes the above clause "MAY fail the connection" and disconnects.

Reply:
We could add an explicit  `reply_channel_range_end` message here but I don't think that it is necessary: you can check the `first_blocknum` and `number_of_blocks` fields and tell when you've received all the replies to your query.

Reply:
It would be good to clarify what rules exactly apply to determine if the last N `reply_channel_range` has fulfilled the most recent `query_channel_range` or not.  Later the text mentions using canned 1000-block `reply_channel_range`s, for instance; presumably a query is covered if all `reply_channel_range`s have fully covered the range, but for instance is it allowed to return the replies in any order, including random?

Reply:
Short channel ids are ordered and replies must be sent in order (but it is not explicit in the current wording when replies span several messages).

Reply:
I've changed the language appropriately:

```diff
 The sender of `query_channel_range`:
-  - MUST NOT send this if it has sent a previous `query_channel_range` to this peer and not received `reply_channel_range`.
+  - MUST NOT send this if it has sent a previous `query_channel_range` to this peer and not received all `reply_channel_range` replies.
```

and

```diff
 The receiver of `query_channel_range`:
-  - if it has not sent `reply_channel_range` to a previously received `query_channel_range` from this sender:
+  - if it has not sent all `reply_channel_range` to a previously received `query_channel_range` from this sender:
     - MAY fail the connection.
```

Since it's made clear that the receiver has to cover the entire range, the definition of `all` seems simple enough to leave implied?